Netstat.pm

```diff
@@ -0,0 +1,243 @@
+
+use strict;
+use warnings;
+
+use parent 'Exporter';
+our @EXPORT= qw(
+    generate_diff_netstat_s
+);
+
+# netstat output might use plural for some nouns always use
+# the plural form of these words.
+sub canonicalize_key {
+    my ($key) = @_;
+    my @plurales = qw(miss);
+    my @pluralys = qw(entr);
+    my @plurals = qw(ACK Interface SYN TDB accept ack agreement allocation
+	association attempt authentication byte calculation call change
+	checksum cleanup collision connection datagram decapsulation
+	decryption destination drop duplicate episode error failure field
+	flow fragment frame gap gateway global insert jump llx local
+	lookup mbuf message mismatche node notification option overflow
+	packet prediction probe quer redirect replay report request
+	response rexmit route scan seed segment slide state table
+	timeout transition upcall use);
+
+    $key =~ s/miss/misses/;
+    $key =~ s/entry/entries/;
+    foreach my $sub (@plurals) {
+	$key =~ s/ ${sub}([^s])/ ${sub}$1/;
+	$key =~ s/^${sub}([^s])/${sub}s$1/;
+    }
+
+    chomp $key;
+    return $key;
+}
+
+# map netstat indentation to perl hash some nestat -s lines increase the
+# indentation and at the same time have a value.  Write that value to the
+# "total" field.
+sub parse {
+    my ($fname) = @_;
+    my %netstat;
+    my $l1;
+    my $l2;
+    my $l3;
+    open(my $fh, '<', "$fname") or die $!;
+
+    while(<$fh>) {
+	if ($_ =~ /^(\w+):$/) {
+	    $l1 = canonicalize_key $1;
+	    $netstat{$l1} = {}
+	} elsif ($_ =~ /^			([^:	]+): (\d+)$/) {
+	    my $k = canonicalize_key($1);
+	    if (ref($netstat{$l1}{$l2}{$l3}) ne "HASH") {
+		my $total = $netstat{$l1}{$l2}{$l3};
+		$netstat{$l1}{$l2}{$l3} = {};
+		$netstat{$l1}{$l2}{$l3}{total} = $total;
+	    }
+	    $netstat{$l1}{$l2}{$l3}{$k} = $2;
+	} elsif ($_ =~ /^		(\d+) (.+)$/) {
+	    $l3 = canonicalize_key $2;
+	    my $v = $1;
+	    if (ref($netstat{$l1}{$l2}) ne "HASH") {
+		my $total = $netstat{$l1}{$l2};
+		$netstat{$l1}{$l2} = {};
+		$netstat{$l1}{$l2}{total} = $total;
+	    }
+	    if ($l3 =~ /\((.+)\)/ && $1 =~ /\d/) {
+		    my ($l) = $l3 =~ /\((.+)\)/;
+		    $l3 =~ s/ \(.+\)//;
+		    (my $k2 = "$l3 $l") =~ s/\d+ //;
+		    $k2 = canonicalize_key $k2;
+		    my ($v2) = $l =~ /(\d+)/;
+		    $netstat{$l1}{$l2}{$k2} = $v2;
+	    }
+	    $netstat{$l1}{$l2}{$l3} = $v;
+	} elsif ($_ =~ /^		([^:]+): (\d+)$/) {
+	    $l3 = canonicalize_key $1;
+	    $netstat{$l1}{$l2}{$l3} = $2;
+	} elsif ($_ =~ /^	(\d+) (.+)$/) {
+	    my $v = $1;
+	    $l2 = canonicalize_key $2;
+	    if ($l2 =~ /\((.+)\)/ && $1 =~ /\d/) {
+		    my ($l) = $l2 =~ /\((.+)\)/;
+		    $l2 =~ s/ \(.+\)//;
+		    (my $k2 = "$l2 $l") =~ s/\d+ //;
+		    $k2 = canonicalize_key $k2;
+		    my ($v2) = $l =~ /(\d+)/;
+		    $netstat{$l1}{$k2} = $v2;
+	    }
+	    $netstat{$l1}{$l2} = $v;
+	} elsif ($_ =~ /^	([^:]+):?$/) {
+	    $l2 = canonicalize_key $1;
+	    $netstat{$l1}{$l2} = {};
+	} else {
+	    die "Can't parse '$_'";
+	}
+    }
+
+    close($fh);
```

Comment:
I explicitly call close() if I want to check the error code.  In this case Perl automatically closes $fh when it goes out of scope.  The line can be omitted.